File: src/python/challenges/problems/valid_palindrome_test.py
# https://leetcode.com/problems/valid-palindrome

from core.problem_base import *

class IsPalindrome(ProblemBase):
    def Solution(self, s: str) -> bool:
        left = 0
        right = len(s) - 1
        while left < right:
            lRune = s[left]
            rRune = s[right]

            if not (lRune.isalpha() or lRune.isdigit()):
                left += 1
                continue

            if not (rRune.isalpha() or rRune.isdigit()):
                right -= 1
                continue

            if lRune.lower() != rRune.lower():
                return False

            left += 1
            right -= 1

        return True

if __name__ == '__main__':
    TestGen(IsPalindrome) \
        .Add(lambda tc: tc.Param("A man, a plan, a canal: Panama").Result(True)) \
        .Add(lambda tc: tc.Param("race a car").Result(False)) \
        .Add(lambda tc: tc.Param(" ").Result(True)) \
        .Run()

File: src/rust/challenges/src/problems/valid_palindrome_test.rs
// https://leetcode.com/problems/valid-palindrome

pub fn solution(s: String) -> bool {
    let mut left = 0;
    let mut right = s.len() - 1;
    let s = s.chars().collect::<Vec<char>>();

    while left < right {
        let l_rune = s[left];
        let r_rune = s[right];

        if !l_rune.is_ascii_alphanumeric() {
            left += 1;
            continue;
        }

        if !r_rune.is_ascii_alphanumeric() {
            right -= 1;
            continue;
        }

        if l_rune.to_ascii_lowercase() != r_rune.to_ascii_lowercase() {
            return false;
        }

        left += 1;
        right -= 1;
    }

    true
}